File: solutions/ocaml/roman-numerals/1/roman_numerals.ml
let roman_numerals = [
    (1000, "M");
    (900,  "CM");
    (500,  "D");
    (400,  "CD");
    (100,  "C");
    (90,   "XC");
    (50,   "L");
    (40,   "XL");
    (10,   "X");
    (9,    "IX");
    (5,    "V");
    (4,    "IV");
    (1,    "I");
]

let rec build_roman remaining value_symbol_pairs =
    match remaining, value_symbol_pairs with
    | 0, _ ->
        ""
    | _, [] ->
        ""
    | value, (symbol_value, symbol) :: rest ->
        if value >= symbol_value
        then
          let remainder = value - symbol_value in
          symbol ^ build_roman remainder value_symbol_pairs
        else
          build_roman value rest

let to_roman number =
  build_roman number roman_numerals

File: solutions/java/roman-numerals/1/src/main/java/RomanNumerals.java
import java.util.Arrays;
import java.util.List;

class RomanNumerals {
    private final int number;

    RomanNumerals(int number) {
        this.number = number;
    }

    private static final List<RomanSymbol> ROMAN_SYMBOLS = Arrays.asList(
            new RomanSymbol("M", 1000),
            new RomanSymbol("CM", 900),
            new RomanSymbol("D", 500),
            new RomanSymbol("CD", 400),
            new RomanSymbol("C", 100),
            new RomanSymbol("XC", 90),
            new RomanSymbol("L", 50),
            new RomanSymbol("XL", 40),
            new RomanSymbol("X", 10),
            new RomanSymbol("IX", 9),
            new RomanSymbol("V", 5),
            new RomanSymbol("IV", 4),
            new RomanSymbol("I", 1)
    );

    String getRomanNumeral() {
        int remaining = number;
        StringBuilder result = new StringBuilder();

        for (RomanSymbol romanSymbol : ROMAN_SYMBOLS) {
            while (remaining >= romanSymbol.value()) {
                result.append(romanSymbol.symbol());
                remaining -= romanSymbol.value();
            }
        }

        return result.toString();
    }
}